packages/SwingSet/src/controller/upgradeSwingset.js

```diff
@@ -204,6 +208,156 @@ export const upgradeSwingset = kernelStorage => {
     version = 2;
   }
 
+  if (version < 3) {
+    // v3 means that we've completed remediation for bug #9039
+    console.log(`Starting remediation of bug #9039`);
+
+    // find all terminated vats
+    const terminated = new Set(JSON.parse(getRequired('vats.terminated')));
+
+    // find all live vats
+    const allVatIDs = [];
+    for (const [_name, vatID] of getAllStaticVats(kvStore)) {
+      if (!terminated.has(vatID)) {
+        allVatIDs.push(vatID);
+      }
+    }
+    for (const vatID of getAllDynamicVats(getRequired)) {
+      if (!terminated.has(vatID)) {
+        allVatIDs.push(vatID);
+      }
+    }
+
+    // find all pending notifies
+    const notifies = new Map(); // .get(kpid) = [vatIDs..];
+    const [runHead, runTail] = JSON.parse(getRequired('runQueue'));
+    for (let p = runHead; p < runTail; p += 1) {
+      const rq = JSON.parse(getRequired(`runQueue.${p}`));
+      if (rq.type === 'notify') {
+        const { vatID, kpid } = rq;
+        assert(vatID);
+        assert(kpid);
+        let vats = notifies.get(kpid);
+        if (!vats) {
+          vats = [];
+          notifies.set(kpid, vats);
+        }
+        vats.push(vatID);
+      }
+    }
+    const [accHead, accTail] = JSON.parse(getRequired('acceptanceQueue'));
+    for (let p = accHead; p < accTail; p += 1) {
+      const rq = JSON.parse(getRequired(`acceptanceQueue.${p}`));
+      if (rq.type === 'notify') {
+        const { vatID, kpid } = rq;
+        assert(vatID);
+        assert(kpid);
+        if (!notifies.has(kpid)) {
+          notifies.set(kpid, []);
+        }
+        notifies.get(kpid).push(vatID);
+      }
+    }
```

Comment:
Better: extract `queueLength`/`dumpQueue` from kernelKeeper.js like you are already doing with `enqueue` → `addToQueue`.
```suggestion
    for (const queueName of ['runQueue', 'acceptanceQueue']) {
      for (const { type, vatID, kpid } of dumpQueue(queueName)) {
        if (type === 'notify') {
          assert(vatID);
          assert(kpid);
          if (!notifies.has(kpid)) {
            notifies.set(kpid, []);
          }
          notifies.get(kpid).push(vatID);
        }
      }
    }
```

Reply:
done